stdlib/REPL/src/LineEdit.jl

```diff
@@ -189,36 +186,31 @@ complete_line(c::CompletionProvider, s, ::Module) = complete_line(c, s)
 terminal(s::IO) = s
 terminal(s::PromptState) = s.terminal
 
-
 function beep(s::PromptState, duration::Real=options(s).beep_duration,
-              blink::Real=options(s).beep_blink,
-              maxduration::Real=options(s).beep_maxduration;
-              colors=options(s).beep_colors,
+    blink::Real=options(s).beep_blink,
+    maxduration::Real=options(s).beep_maxduration;
+              beep_face=options(s).beep_face,
               use_current::Bool=options(s).beep_use_current)
     isinteractive() || return # some tests fail on some platforms
     s.beeping = min(s.beeping + duration, maxduration)
-    let colors = Base.copymutable(colors)
-        errormonitor(@async begin
-            trylock(s.refresh_lock) || return
-            try
-                orig_prefix = s.p.prompt_prefix
-                use_current && push!(colors, prompt_string(orig_prefix))
-                i = 0
-                while s.beeping > 0.0
-                    prefix = colors[mod1(i+=1, end)]
-                    s.p.prompt_prefix = prefix
-                    refresh_multi_line(s, beeping=true)
-                    sleep(blink)
-                    s.beeping -= blink
-                end
-                s.p.prompt_prefix = orig_prefix
-                refresh_multi_line(s, beeping=true)
-                s.beeping = 0.0
-            finally
-                unlock(s.refresh_lock)
+    errormonitor(@async begin
+        trylock(s.refresh_lock) || return
+        try
+            og_prompt = s.p.prompt
+            beep_prompt = styled"{$beep_face:$(prompt_string(og_prompt))}"
```

Comment:
@tecosaur : do you have feedback on this approach, or guidance for a better one? For context, the problem here is that the prompt string needs to be styled with the `repl_prompt_beep` face for a short period (to indicate a non-action to the user, like when pressing backspace at the beginning of a prompt)

I initially tried to use `withfaces`, but realized that doing so would require overriding all prompt faces (`:repl_prompt_*`), and doing so would:
1) make future modifications slightly more fragile and
2) exclude user-defined REPL modes.

Thus, my solution here is to create a new styled string which wraps the prompt string in a new style, `$beep_face`, and then restores the original prompt afterward.

Reply:
Faces applied later take priority, so applying `repl_prompt_face` around should do the trick nicely I think :slightly_smiling_face:

![image](https://github.com/JuliaLang/julia/assets/20903656/206a139b-9b36-42d8-97d1-456107a19646)

I'm curious though, is there any particular reason why `beep_face` is a variable and you don't just use `repl_prompt_beep`?

Reply:
Great, thank you for the direction :+1:

> is there any particular reason why `beep_face` is a variable and you don't just use `repl_prompt_beep`

I think my original intent was to maintain the pre-existing logic for the case when color was disabled. Essentially what was removed here: 

https://github.com/JuliaLang/julia/pull/51887/files#diff-730f90e0bf9ca018477a1d4a52e0d7398af2bb8cf9e401568fd86690eec81bb7L499

You can see in that section of the diff that it's removed and hasn't been actually re-implemented, but perhaps a variable like this isn't needed at all, now that StyledStrings is used? In other words, the original variable existed so that the terminal's color state could be carried through to the point of printing the beep control sequences (or not), but is no longer necessary (I think?) because it is precisely the problem which StyledStrings solves. Does that sound right @tecosaur?

Reply:
I think so :grinning:.

If we're interpreting this correctly, a chunk of this code was originally needed because of the lack of composable styling, but that's a feature that `StyledStrings` has OOTB.

That said, there's also the complication that it seems like `beep_colors` is designed so that it could hypothetically cycle through a rainbow of colours? This seems rather strange though. Similarly, there are other aspects that just don't make sense to me, like `beep_use_current`. I don't understand when you'd want it to be false, and I went through all 104 results for that symbol in GitHub's site-wide search, and couldn't come across any code that set it to anything but `true` :confused:.

The first (of two) commit I could find with it was 176796360fa704b7e340d2cbd56a80d87568014e (6 years ago), and that's not introducing it but replacing a global `const BEEP_USE_CURRENT = Ref(true)`. That const itself came from 8c2fc37ac939a3f30de4744ab0fa6b6803a7a4b3 just a few months earlier, and the PR that introduces it has this for explanation:

> I like it sober, so it blinks once by default, alternating with :light_black color (can be customized).

I think with the benefit of hindsight I'm inclined to view the addition of five variables to control the behaviour of the prompt "beep" as an enthusiastic addition which isn't retrospectively worth the complication it adds, given I haven't heard of anyone customising it previously, nor have I been able to find any examples by searching GitHub.

@rfourquet perhaps you might have something to add here?

Reply:
If we're able to reduce the number of beep customisations, we could potentially reduce the main logic to just something like this (untested):

```julia
try
    orig_prompt = s.p.prompt
    prompt_str = prompt_string(orig_prompt)
    isbeep = false
    while s.beeping > 0.0
        s.p.prompt = if (isbeep = !isbeep)
            styled"{repl_beep:$prompt_str}"
        else prompt_str end
        refresh_multi_line(s, beeping=true)
        sleep(blink)
        s.beeping -= blink
    end
    s.p.prompt = orig_prompt
    refresh_multi_line(s, beeping=true)
    s.beeping = 0.0
finally
    unlock(s.refresh_lock)
end
```